### DISCUSSION TITLE
FastAPI does not honour validator pre=True

### DISCUSSION THREAD

**Discussion by michaelspori** [fastapi/fastapi#9853] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import Depends, FastAPI, Request
from pydantic import BaseModel, validator

class Params(BaseModel):
    ids: list[str] = []

    @validator("ids", pre=True)
    def split_string(cls, val):
        print("validator: received", val.__repr__())
        return val.split(",") if isinstance(val, str) else ["boo"]

app = FastAPI()

@app.get("/test")
def test(request: Request, params: Params = Depends()):
    print(">> via fastapi:", params)
    print(">> directly:", Params(**dict(request.query_params)))
    return params
```

### Description

Run the above code with `uvicorn myfile:app --reload`, where `myfile` is the file name, and hit it with `curl http://127.0.0.1:8000/test?ids=1,2,3`

Server output:
```
validator: received []
>> via fastapi: ids=['boo']
validator: received '1,2,3'
>> directly: ids=['1', '2', '3']
```

The validator should run *before* `ids` is converted to a list, because I set `pre=True`.

If I construct the model manually, as per the `>> directly` line, you can see the value received by the validator is the `"1,2,3"` string. I can then split it within the validator to create the list which the field expects. This is the correct behaviour.

However, when fastapi constructed the `params` object for me, the validator was executed but only when the value had already become an empty list, as per "validator: received []". I believe this is incorrect behaviour because it means I can no longer use the validator to do something like a string split, and the original value is already "lost" as far as constructing the model is concerned.

The root validator has the same behaviour. I also tried overriding `__new__` and `__init__` and printing out the args to see if I could catch them there, but the type coercion/data loss has already occurred at that point.

I'm aware of some workarounds (e.g. using a method to construct the class) but there are various other downsides to them. I'd much rather get it working purely via `params: Params = Depends()` if possible.

### Operating System

Windows

### FastAPI Version

0.85.1

### Python Version

3.10.8

### Additional Context

Pydantic 1.10.7

**Comment by ahmedabdou14** [comment]
I think you miss interpreted the outputs. Pydantic is working completely as it should.

The reason why it defaulted to empty list at first is because it was never initialized inside the request body so it never triggered the validator. Since it was not assigned a value,  it will be defaulted to empty list which triggers the validator that then converts `ids` into `["boo"]`

basically this took place when the request body was empty
```python
def test(request: Request, params = Params(ids=[])):
```

It is like the default value is injected during initialization and not after.

_Side note:_
Having an empty list `[]` as a default value would make all instances of Params to reference the same list for `ids`, so changing one would change all other instances. Instead use:
```python
class Params(BaseModel):
    ids: list[str] = Field(default_factory=list)
```

**Reply by michaelspori** [reply]
I'm sure Pydantic is working correctly, and my interpretation of what's going on is evidently wrong. However that doesn't change the issue: `Params` should *not* be receiving the empty list as default because there *is* an `ids` query parameter whose value should be supplied instead. If I change `ids` in the model to `str`, then it receives that value. So something (fastapi, I assume) is deciding to *not* supply the ids string as the parameter value when `ids` is a list, probably because it can't implicitly convert the type, but this prevents me from using the validator to convert it to the correct type - as is done when I instantiate the model manually.

Thanks for reminding me about the empty list default, I'm still new to python so forget about these things.

-------

Update: you mentioned the request body being empty, which made me wonder if it was actually trying to get the value from the body, and of course it wouldn't be there, in which case the default would be used - sure enough that's what was happening. I changed the model to:
```
ids: list[str] = Field(Query())
```
And now it's actually getting the value from the query parameters, resulting in `ids` of `["1,2,3"]`.

It's still not quite right however - I should be getting `"1,2,3"` not `["1,2,3"]`. Evidently there is some preprocessing going on before the values are supplied to my model, which is a problem for me because if `ids` is typed as `list[int]` (which it is in my actual use case) then I get a 422 error with a message about it not being a valid integer (validator was changed to return list of int). But when constructing manually, with `ids="1,2,3"`, it works fine because the validator converts them to a list of int, and there is nothing checking and failing beforehand.

**Reply by ahmedabdou14** [reply]
Sorry @michaelspori but there are alot of python/fastapi/pydantic concepts that you are missing here and it is very tough to explain to you all the stuff you are missing via github discussions. The reason why you get an error when you pass in `"1,2,3"` to a type of `list[int]` is because the validator just splits the string into a list of strings `['1', '2', '3']` which is of type `list[str]` not `list[int]`. You can fix it by casting each element to `int` inside your validator using this
```python
@validator("ids", pre=True)
def split_string(cls, val):
    print("validator: received", val.__repr__())
    return [int(x) for x in string.split(",")] if isinstance(val, str) else ["boo"]
```

_Another sidenote:_
```python
# this is not the correct way to define a query parameter
ids: list[str] = Field(Query())
# Instead use this
ids: list[str] = Query(default_factory=list)
```

**Reply by michaelspori** [reply]
No, I understand that. I appreciate your help. However I still feel something is being lost in translation here. Maybe on my side. I didn't post the code but I mentioned that I'd updated the validator to return a list of ints. I did basically what you did in the validator, and I also tried hardcoding it to just return something like `[1,2]` and completely ignore the input. In either case, I got an error saying it's not a valid integer.

To be super clear: If I run this code:
```
from fastapi import Depends, FastAPI, Request, Query
from pydantic import BaseModel, validator, Field


class Params(BaseModel):
    ids: list[int] = Field(Query(default_factory=list))

    @validator("ids", pre=True)
    def split_string(cls, val):
        print("validator: received", val.__repr__())
        return [4, 5]


app = FastAPI()


@app.get("/test")
def test(params: Params = Depends()):
    print(params)
    return params
```
And hit it with this:
```
curl http://127.0.0.1:8000/test?ids=1,2,3
```
I get this error:
```
{"detail":[{"loc":["query","ids",0],"msg":"value is not a valid integer","type":"type_error.integer"}]}
```
And nothing is printed from the validator.

If I change the model to `ids: list[int] = Query(default_factory=list)` then the validator functions receives `None`.
If I change the model to `ids: list[str] = Query(default_factory=list)` then the validator function receives `None`.
If I change the model to `ids: list[str] = Field(Query(default_factory=list))` then the validator function receives `['1,2,3']`.

So none of those seem right. Am I doing something wrong? What I expect is the same behaviour as when I construct the model manually, i.e. the validator receives the "1,2,3" string and spits out a list of ints.

If it's too much of a bother, don't worry, I have a workaround, but it's bothering me that I can't just get it working with validators as I think it should.

**Reply by ahmedabdou14** [reply]
Ok now it is much clearer!

```python
# For some reason, the only was i found to read  a Query parameter as a list from a pydantic model is what you have done 
ids: list[str] = Field(Query(default_factory=list))
# however this is weird because a query parameter should not take in a list, instead a string as you have done
# or else swagger doc will be very weird and ask to input a list of ids like this `ids=1&ids=2&ids=3` which works but it looks very bad imo.

# I do not know why this does not read from query but instead expects ids to be in the request `body`!
ids: list[int] = Query(default_factory=list)
# however this works as expected
ids: str = Query()

# so i guess if it was a list, it will force it into the body of the api
```
I am not sure if it is a bug in pydantic or fastapi. @Kludex can you please investigate this behavior?

Not only that but @michaelspori you are right! it seems like the fastapi validates the query param type before executing the validator. This highlights the fact that validators work before or after assignment, but fastapi will validate the input even before creating the model.

**Comment by sarangj** [comment]
@michaelspori experiencing something similar, did you ever get a resolution here?

**Reply by michaelspori** [reply]
I did not, and still have the same issue as well. I typically create a model for input parameters, so what I do as a workaround is type the value as something that fastapi won't reject (for example, `str`), then create a private attribute of the actual type I want, a public property for that private attribute, and a field or model validator that performs the actual parsing from the fastapi-approved type to the type I want and sets the property. I then use that property throughout the codebase when interacting with the model, rather than the field for the actual value received.

For example:
```python
from pydantic import BaseModel, model_validator, PrivateAttr

class SampleParams(BaseModel):
    ids: str | None = None
    _entity_ids: list[int] | None = PrivateAttr(default=None)

    @property
    def entity_ids(self):
        return self._entity_ids
    
    @entity_ids.setter
    def entity_ids(self, val: list[int] | None):
        self._entity_ids = val
    
    @model_validator
    def validate_model(self):
        if self.ids:
            ids = [x.strip() for x in self.ids.split(",")]
            self.entity_ids = [int(x) for x in ids if x]
```
Then I only ever access `entity_ids` rather than `ids`.

**Reply by sarangj** [reply]
Got it, thanks for the example @michaelspori! This is essentially what I ended up doing as well (though TIL `PrivateAttr`, I couldn't get a field with a leading underscore to work with Pydantic otherwise).  A little frustrating that FastAPI doesn't seem to respect Pydantic validation, but :shrug:

**Comment by YuriiMotov** [comment, marked as answer]
Using Pydantic `BaseModel` with `Depends` was never officially supported in FastAPI. It partially works, but not always as expected.

Since 0.115.0 FastAPI you can officially [use Pydantic models to declare Query parameters](https://fastapi.tiangolo.com/tutorial/query-param-models/).
To do this, you should replace `params: Params = Depends()` with `params: Params = Query()`.
You should also modify your validator to expect list of strings, not just string - since parameter is annotated as `list`, FastAPI will extract it as list and pass to your validator.

```py
from fastapi import Query, FastAPI, Request
from pydantic import BaseModel, field_validator


class Params(BaseModel):
    ids: list[str] = []

    @field_validator("ids", mode="before")
    @classmethod
    def split_string(cls, val):
        print("validator: received", val.__repr__())
        if isinstance(val, list) and len(val) == 1 and isinstance(val[0], str):
            val = val[0]
        return val.split(",") if isinstance(val, str) else ["boo"]

app = FastAPI()

@app.get("/test")
def test(request: Request, params: Params = Query()):
    print(">> via fastapi:", params)
    print(">> directly:", Params(**dict(request.query_params)))
    return params

```

`curl http://127.0.0.1:8000/test?ids=1,2,3`

Output:

```
validator: received ['1,2,3']
>> via fastapi: ids=['1', '2', '3']
validator: received '1,2,3'
>> directly: ids=['1', '2', '3']
```

---

There is currently a limitation with this approach - you can't use multiple Query parameter models and you can't combine Query parameter model with single Query parameters in one endpoint.
In some cases you can overcome it by using multiple inheritance.
There is a PR to implement the support for multiple parameter models: https://github.com/fastapi/fastapi/pull/12481
